apps/src/blockly/addons/cdoUtils.ts

```diff
@@ -503,6 +529,44 @@ export function getLevelToolboxBlocks(customCategory: string) {
   }
 }
 
+/**
+ * Creates a map of block types and limits, based on limit attributes found
+ * in the block XML for the current toolbox.
+ * @returns {Map<string, number>} A map of block limits
+ */
+export function createBlockLimitMap() {
+  const parser = new DOMParser();
+  // This method only works for string toolboxes.
+  if (!Blockly.toolboxBlocks || typeof Blockly.toolboxBlocks !== 'string') {
+    return;
+  }
+
+  const xmlDoc = parser.parseFromString(Blockly.toolboxBlocks, 'text/xml');
+  // Define blockLimitMap
+  const blockLimitMap = new Map<string, number>();
+
+  // Select all block elements and convert NodeList to array
+  const blocks = Array.from(xmlDoc.querySelectorAll('block'));
+
+  // Iterate over each block element using forEach
+  blocks.forEach(block => {
+    // Check if the block has a limit attribute
+    const limitAttribute = block.getAttribute('limit');
+    if (limitAttribute !== null) {
+      // Check if the limit attribute is a valid number
+      const limit = parseInt(limitAttribute);
+      if (!isNaN(limit)) {
+        // Extract type and add to blockLimitMap
+        const type = block.getAttribute('type');
```

Comment:
If `getAttribute` returns an empty string, do you want to call `blockLimitMap.set`?

Reply:
I don't think it can? All block element should have a type. The only way around that would be to manually add `<block type=""/>` to the level XML, but I don't think we need to be overly concerned about a graceful failure in a case like that.

Reply:
Sounds good. I based this conversation on MDN, https://developer.mozilla.org/en-US/docs/Web/API/Element/getAttribute, but this is blocks not `Element`s!

Reply:
The blocks are xml elements (renaming to make that clear). The documentation you linked says:

> All modern web browsers return `null` when the specified attribute does not exist on the specified element.

This is why I do a null check below. I think we may be overanalyzing this though because Blockly guarantees that all block elements will have a type.